setup.py

```diff
@@ -780,76 +826,135 @@ def get_provider_package_from_package_id(package_id: str):
+        # And it results with an error when running pip directly (cannot import pip module)
+        # also PIP does not have a stable API so we have to run subprocesses ¯\_(ツ)_/¯
+        try:
+            installed_packages = (
+                subprocess.check_output(["python3", "-m", "pip", "freeze"]).decode().splitlines()
+            )
+            airflow_provider_packages = [
+                package_line.split("=")[0]
+                for package_line in installed_packages
+                if package_line.startswith("apache-airflow-providers")
+            ]
+            self.announce(f'Uninstalling ${airflow_provider_packages}!', level=log.INFO)
+            subprocess.check_call(["python3", "-m", "pip", "uninstall", "--yes", *airflow_provider_packages])
+        except subprocess.CalledProcessError as e:
+            self.announce(f'Error when uninstalling airflow provider packages: {e}!', level=log.WARN)
+        super().run()
+
+
+class Install(install_orig):
+    """Forces installation of providers from sources in editable mode."""
 
+    def run(self):
+        self.announce('Standard installation. Providers are installed from packages', level=log.INFO)
+        super().run()
 
-def do_setup():
-    """Perform the Airflow package setup."""
+
+def do_setup() -> None:
+    """
+    Perform the Airflow package setup.
+    Most values come from setup.cfg, only the dynamically calculated ones are passed to setup
+    function call. See https://setuptools.readthedocs.io/en/latest/userguide/declarative_config.html
+    """
     setup_kwargs = {}
 
-    if os.getenv('INSTALL_PROVIDERS_FROM_SOURCES') == 'true':
-        # Only specify this if we need this option, otherwise let default from
-        # setup.cfg control this (kwargs in setup() call take priority)
-        setup_kwargs['packages'] = find_namespace_packages(include=['airflow*'])
+    def include_provider_namespace_packages_when_installing_from_sources() -> None:
+        """
+        When install providers from sources we install all namespace packages found below airflow,
+        including airflow and provider packages, otherwise defaults from setup.cfg control this.
+        The kwargs in setup() call override those that are specified in setup.cfg.
+        """
+        if os.getenv(INSTALL_PROVIDERS_FROM_SOURCES) == 'true':
+            setup_kwargs['packages'] = find_namespace_packages(include=['airflow*'])
+
+    include_provider_namespace_packages_when_installing_from_sources()
+    if os.getenv(INSTALL_PROVIDERS_FROM_SOURCES) == 'true':
+        print("Installing providers from sources. Skip adding providers as dependencies")
     else:
         add_all_provider_packages()
+
     write_version()
     setup(
         distclass=AirflowDistribution,
-        # Most values come from setup.cfg -- see
-        # https://setuptools.readthedocs.io/en/latest/userguide/declarative_config.html
```

Comment:
Did you mean to remove this comment?

Reply:
Yep. I rephrased it and put a bit more comprehensive information (with dynamic vs. static content) few lines above or below.